docs/rfcs/20220829-distributed-dynamic-routing.md

```diff
@@ -0,0 +1,224 @@
+- CeresMeta maintains a Map<Table, Tablecost> for each computing node. When the load of a node reaches the threshold, rebalance is triggered, some tables are selected, migrated to the computing node with the lowest current load, and the routing relationship between shard and table is adjusted.
+```
+    ┌──────┐        ┌──────┐         ┌──────┐    ┌───────┐    ┌──────┐        ┌──────┐         ┌──────┐    
+ ┌──┤Node0 ├──┐  ┌──┤Node1 ├──┐   ┌──┤Node2 ├──┐ │       │ ┌──┤Node0 ├──┐  ┌──┤Node1 ├──┐   ┌──┤Node2 ├──┐ 
+ │  └──────┘  │  │  └──────┘  │   │  └──────┘  │ │ Table │ │  └──────┘  │  │  └──────┘  │   │  └──────┘  │ 
+ │┌────────┐  │  │            │   │┌────────┐  │ │ Load  │ │┌────────┐  │  │            │   │┌────────┐  │ 
+ ││Shard0 L├┐ │  │┌─────────┐ │   ││Shard1 L├┐ │ │Balance│ ││Shard0 L├┐ │  │┌─────────┐ │   ││Shard1 L├┐ │ 
+ │├────────┘│ │  ││Shard0 F │ │   │├────────┘│ │ │       │ │├────────┘│ │  ││Shard0 F │ │   │├────────┘│ │ 
+ ││ table0  │ │  │└─────────┘ │   ││ table2  │ │ └───────┘ ││ table0  │ │  │└─────────┘ │   ││ table1  │ │ 
+ ││ table1  │ │  │┌─────────┐ │   ││ table3  │ │──────────▶││         │ │  │┌─────────┐ │   ││ table2  │ │ 
+ ││         │ │  ││Shard1 F │ │   ││         │ │           ││         │ │  ││Shard1 F │ │   ││ table3  │ │ 
+ │└─────────┘ │  │└─────────┘ │   │└─────────┘ │           │└─────────┘ │  │└─────────┘ │   │└─────────┘ │ 
+ │┌─────────┐ │  │            │   │┌─────────┐ │           │┌─────────┐ │  │            │   │┌─────────┐ │ 
+ ││Shard2 L │ │  │            │   ││Shard2 F │ │           ││Shard2 L │ │  │            │   ││Shard2 F │ │ 
+ │└─────────┘ │  │            │   │└─────────┘ │           │└─────────┘ │  │            │   │└─────────┘ │ 
+┌┴────────────┴┐ └────────────┘  ┌┴────────────┴┐         ┌┴────────────┴┐ └────────────┘  ┌┴────────────┴┐
+│  CPUUse=90%  │                 │  CPUUse=30%  │         │  CPUUse=50%  │                 │  CPUUse=50%  │
+└──────────────┘                 └──────────────┘         └──────────────┘                 └──────────────┘
+```
+
+### Shard Change Process
+#### Switch Leader
+
+1. First, set the original leader shard and the follwer to be switched to PENDING_ FOLLOWER and PENDING_ LEADER.
+1. Wait for the leader shard to complete the preparatory action before the leader switch, and notify the CeresMeta after the completion. In this process, the new leader cannot process the write request.
+1. After receiving this request, meta will update the status of the two shards, and the leader switch is completed.
+
+
+#### Shard Migrate
+
+- Leader Shard：
+    1. First, CeresMeta adjusts the routing relationship and create a new follower shard on the node.
+    1. Follower shard notifies CeresMeta after initialization.
+    1. CeresMeta started the process of leader switch, switched the leader to the newly created follower shard.
+    1. Close the original leader shard after the leader switch is completed.
+- Follower Shard：
```

Comment:
For the case where replication_factor > 1, the follower shard may follow the same procedure as that when replication_factor == 1:
> 1. Close the shard on the old shard;
> 2. Open the shard on the new shard;
> 3. Adjust shard node mapping;